File: C++/ParenthesesBalance.cpp
/** https://uva.onlinejudge.org/index.php?option=onlinejudge&page=show_problem&problem=614
 * idea: use stack
 */
#include<iostream>
#include<string>
#include<deque>

using namespace std;

char validPairChar(char c) {
    if (c == ']') {
        return '[';
    } else if (c == ')') {
        return '(';
    } else {
        return c;
    }
}
string check(string& str) {
    deque<char> stack;
    
    for (string::iterator it = str.begin(); it != str.end(); it++) {
        char c = *it;
#ifdef _BETTER
        if (stack.empty()) {
            stack.push_front(c);
        } else {
            if (c != validPairChar(c)) {
                if (stack.front() != validPairChar(c)) {
                    return "No";
                } else {
                    stack.pop_front();
                }
            } else {
                stack.push_front(c);
            }
        }
#else _BETTER
 // general way
        if (c == '(' || c == '[') 
            stack.push_front(c);
        else { // close
            if (stack.empty() || validPairChar(c) != stack.front()) {
                return "No";
            } else {
                stack.pop_front();
            }
        }
#endif  // _BETTER
    }
    if (!stack.empty()) {
        return "No";
    }
    return "Yes";
}

int main() {
    string str;
    int n;
    cin >> n;
    getline(cin, str);
    string arr[n];
    for (int i=0; i < n; i++) {
        getline(cin, str);
        arr[i] = str;
        
    }
    for(string& str : arr) {
        cout << check(str) << endl;
    }
    return 0;
}

File: Java/ParenthesesBalance.java
/** https://uva.onlinejudge.org/index.php?option=onlinejudge&page=show_problem&problem=614
 * idea: use stack
 */
import java.util.Scanner;
import java.util.Deque;
import java.util.LinkedList;

class ParenthesesBalance {
    static Character validPairChar(Character c) {
        if (c == ']') {
            return '[';
        } else if (c == ')') {
            return '(';
        } else {
            return c;
        }
    }
    static String check(String str) {
        Deque<Character> stack = new LinkedList<>();
        for (Character c: str.toCharArray()) {
            if (c == '(' || c == '[') {
                stack.addFirst(c);
            } else {
                if (stack.isEmpty() || stack.peekFirst() != validPairChar(c)) {
                    return "No";
                } else {
                    stack.pollFirst();
                }
            }
        }
        if (!stack.isEmpty()) {
            return "No";
        }
        return "Yes";
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        //sc.nextLine();
        
        String[] arr = new String[n];
        for (int i=0; i < n; i++) {
            String str = sc.next(); //sc.nextLine();
            arr[i] = str;
        }
        for (String str: arr) {
            System.out.println(check(str));
        }

    }
}